### ISSUE TITLE
New rule: Don't pass default values to functions

### ISSUE THREAD

**Issue by fregante** [typescript-eslint/typescript-eslint#2008]
This rule would warn using function parameters that are already set by the function's defaults.

Specifying default values for a function adds extra code that needs to be reviewed and nobody likes that. 

Why write:

```js
fetch('/contacts', {
	method: 'GET',
	headers: {},
	mode: 'cors',
	credentials: 'include',
	cache: 'default',
	redirect: 'follow'
});
```

when it means the same as:

```js
fetch('/contacts');
```

### Further examples

```js
function run(always = true) {}
run(true); // Error
```
```js
function run({always} = {always: true}) {}
run({always: true}); // Error
```
```js
function run({always = true} = {}) {}
run({always: true}); // Error
```
```js
function run(options) {
	options = {
		always: true,
		...options
	}
}
run({always: true}); // Error
```
```js
function run(options) {
	options = Object.assign({
		always: true
	}, options);
}
run({always: true}); // Error
```

And any other common defaults pattern (unless already blocked by other rules like [`prefer-object-spread`](https://eslint.org/docs/rules/prefer-object-spread))

**Comment by bradzacher** [comment]
Note: your last few examples where the `options` object is overridden in the function body are getting into the "impossible to detect" area.

Simple cases are easy, sure, but they very quickly become exponentially hard to reliably handle.

I believe it would be trivial to handle if the defaults are set inline, however.

**Comment by fregante** [comment]
> are getting into the "impossible to detect" area.

I don't doubt it. Inline defaults can be a start and the rest of them would be "nice to have"

**Comment by fregante** [comment]
I don’t think this is currently possible. More info on https://github.com/eslint/eslint/issues/13168
